ssds/blobstore/s3.py

```diff
@@ -65,11 +68,19 @@ def __iter__(self):
         if 1 == self.number_of_parts:
             yield self._get_part(0)
         else:
-            futures = list()
-            for part_number in range(self.number_of_parts):
-                futures.append(self._executor.submit(self._get_part, part_number))
-            for f in as_completed(futures):
-                yield f.result()
+            futures: Set[Future] = set()
+            part_numbers = [part_number for part_number in range(self.number_of_parts)]
+            while part_numbers or futures:
+                if len(futures) < self.parts_to_buffer:
+                    number_of_parts_to_fetch = self.parts_to_buffer - len(futures)
+                    for part_number in part_numbers[:number_of_parts_to_fetch]:
+                        futures.add(self._executor.submit(self._get_part, part_number))
+                    part_numbers = part_numbers[number_of_parts_to_fetch:]
+                for f in as_completed(futures):
+                    part_number, data = f.result()
+                    futures.remove(f)
+                    yield part_number, data
+                    break
```

Comment:
Should this `break` still be here?  I'd expect `while part_numbers or futures` to be sufficient, but maybe I'm misunderstanding something here.

Reply:
`as_completed` blocks if any of `futures` is not complete. If we don't break out of the inner loop as soon as the first future is available, we make the client wait for the next future to complete before yielding. This would defeat the purpose of asynchronous execution.

Can you think of a way to make the logic more clear? In the meantime, I'll add a comment.

Reply:
That makes sense.  I think adding a comment is sufficient, though the var "futures" was a bit vague to me as to what it represents.

Reply:
Good point. I'll make a issue to refactor for clarity in a future PR.